src/H5Tnative.c

```diff
@@ -739,17 +741,39 @@ H5T__get_native_float(size_t size, H5T_direction_t direction, size_t *struct_ali
 
     FUNC_ENTER_PACKAGE
 
-    assert(size > 0);
+    if (0 == (size = H5T_get_size(dtype)))
+        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, NULL, "not a valid size");
 
     if (direction == H5T_DIR_DEFAULT || direction == H5T_DIR_ASCEND) {
+        if (size <= 2) {
 #ifdef H5_HAVE__FLOAT16
-        if (size <= sizeof(H5__Float16)) {
-            match       = H5T_NATIVE_FLOAT_MATCH_FLOAT16;
-            native_size = sizeof(H5__Float16);
-        }
-        else
+            H5T_t *bf16_le_dt = NULL; /* Datatype for little-endian bfloat16 */
+            H5T_t *bf16_be_dt = NULL; /* Datatype for big-endian bfloat16 */
+
+            if (NULL == (bf16_le_dt = H5I_object(H5T_FLT_BF16LE)))
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, NULL, "not a data type");
+            if (NULL == (bf16_be_dt = H5I_object(H5T_FLT_BF16BE)))
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, NULL, "not a data type");
+
+            /* Promote bfloat16 to float instead of float16, as it
```

Comment:
bfloat16 types should be promoted to `float` instead of float16, as the type is the same size as float16, but a different format. Converting between bfloat16 and `float` is also very simple (by design).

Reply:
Yes, float should be the closest native type that can represent data correctly and bfloat16's design makes conversion to/from IEEE 32-bit floats simple. This should be revisited when support for a native bfloat16 type is added though, as there would be no need for promotion at that point. I suppose that, yes, there are some formats you could choose for a 16-bit floating point number where this logic would promote them to the _Float16 format instead, which would be problematic. Happy to fix this to just promote everything of this size that isn't _Float16 to float or even do some more format checking to decide the right type to promote to, but there are several other places in the library like that which I believe we should address as well.